File: scripts/extract-2025-data.py
#!/usr/bin/env python3

import openpyxl
import os
import json
import re
from collections import defaultdict

STATS_DIR = "/Users/emir_mw/stats/STATS/2025/Dnevni izvještaji"

airlines = defaultdict(int)
aircraft_types = defaultdict(int)
operation_types = defaultdict(int)

# Valid aircraft type patterns
VALID_AIRCRAFT = re.compile(r'^[A-Z0-9]{2,6}$')
INVALID_AIRCRAFT = re.compile(r'-TZL-|TZL-|^\d{1,3}$')

def normalize_airline(name):
    """Normalize airline name"""
    return name.strip().upper().replace('  ', ' ')

def is_valid_aircraft(aircraft_type):
    """Check if aircraft type is valid"""
    if not aircraft_type:
        return False

    clean = str(aircraft_type).strip()

    # Skip if looks like route
    if INVALID_AIRCRAFT.search(clean):
        return False

    # Skip if too long (probably route)
    if len(clean) > 10:
        return False

    # Valid aircraft pattern
    return VALID_AIRCRAFT.match(clean) is not None

# Parse all months
month_folders = sorted([
    f for f in os.listdir(STATS_DIR)
    if os.path.isdir(os.path.join(STATS_DIR, f))
])

print(f"📂 Found {len(month_folders)} months\n")

total_flights_all = 0

for month_folder in month_folders:
    month_path = os.path.join(STATS_DIR, month_folder)

    # Find Excel file
    files = [
        f for f in os.listdir(month_path)
        if 'Dnevni izvještaj o saobraćaju' in f and f.endswith('.xlsx')
    ]

    if not files:
        continue

    file_path = os.path.join(month_path, files[0])

    try:
        print(f"📊 {month_folder[:10]}: ", end="")

        wb = openpyxl.load_workbook(file_path, read_only=True, data_only=True)
        total_flights = 0

        # Iterate through all sheets (days)
        for sheet_name in wb.sheetnames:
            ws = wb[sheet_name]

            for i, row in enumerate(ws.iter_rows(values_only=True), 1):
                # Skip header
                if i == 1:
                    continue

                # Skip empty rows
                if not row[0]:
                    continue

                total_flights += 1

                # Extract data
                # [0] datum, [1] kompanija, [2] ruta, [3] tip a/c, [4] reg, [5] tip OPER
                airline = row[1]
                aircraft = row[3]
                op_type = row[5]

                # Airline
                if airline and isinstance(airline, str):
                    normalized = normalize_airline(airline)
                    airlines[normalized] += 1

                # Aircraft type
                if aircraft:
                    clean = str(aircraft).strip()
                    if is_valid_aircraft(clean):
                        aircraft_types[clean] += 1

                # Operation type
                if op_type:
                    clean = str(op_type).strip().upper()
                    # Skip if it's a number (capacity)
                    if not re.match(r'^\d+$', clean):
                        operation_types[clean] += 1

        total_flights_all += total_flights
        print(f"✅ {total_flights} flights")
        wb.close()

    except Exception as e:
        print(f"❌ Error: {e}")
        continue

# Sort by count
sorted_airlines = sorted(airlines.items(), key=lambda x: x[1], reverse=True)
sorted_aircraft = sorted(aircraft_types.items(), key=lambda x: x[1], reverse=True)
sorted_op_types = sorted(operation_types.items(), key=lambda x: x[1], reverse=True)

# Print results
print('\n' + '='*80)
print('📊 EKSTRAKTOVANI PODACI IZ 2025 GODINE (CLEAN)')
print('='*80)
print(f'Total flights processed: {total_flights_all}')

print(f'\n✈️  AVIOKOMPANIJE ({len(sorted_airlines)}):')
for name, count in sorted_airlines:
    print(f'  {count:4d} x {name}')

print(f'\n🛩️  TIPOVI AVIONA ({len(sorted_aircraft)}):')
for model, count in sorted_aircraft:
    print(f'  {count:4d} x {model}')

print(f'\n🔧 TIPOVI OPERACIJA ({len(sorted_op_types)}):')
for op_type, count in sorted_op_types:
    print(f'  {count:4d} x {op_type}')

# Save to JSON
output = {
    'airlines': [{'name': name, 'count': count} for name, count in sorted_airlines],
    'aircraftTypes': [{'model': model, 'count': count} for model, count in sorted_aircraft],
    'operationTypes': [{'type': op_type, 'count': count} for op_type, count in sorted_op_types],
    'totalFlights': total_flights_all,
}

os.makedirs('output', exist_ok=True)
with open('output/2025-extracted-data.json', 'w', encoding='utf-8') as f:
    json.dump(output, f, indent=2, ensure_ascii=False)

print('\n✅ Data saved to: output/2025-extracted-data.json')

File: scripts/extract-2025-data.ts
#!/usr/bin/env ts-node

import * as XLSX from 'xlsx';
import * as fs from 'fs';
import * as path from 'path';

const STATS_DIR = '/Users/emir_mw/stats/STATS/2025/Dnevni izvještaji';

interface AirlineData {
  name: string;
  count: number;
}

interface AircraftData {
  model: string;
  count: number;
}

const airlines = new Map<string, number>();
const aircraftTypes = new Map<string, number>();
const operationTypes = new Map<string, number>();

// Valid aircraft type patterns
const VALID_AIRCRAFT = /^[A-Z0-9]{2,6}$/;  // A320, B737, BE40, PC12, etc.
const INVALID_AIRCRAFT = /-TZL-|TZL-|^\d{1,3}$/;  // Rute ili brojevi

// Normalize airline name
function normalizeAirline(name: string): string {
  return name
    .trim()
    .toUpperCase()
    .replace(/\s+/g, ' ');  // Multiple spaces to single
}

// Check if aircraft type is valid
function isValidAircraft(type: string): boolean {
  if (!type) return false;
  const clean = type.trim();

  // Skip if looks like route
  if (INVALID_AIRCRAFT.test(clean)) return false;

  // Skip if too long (probably route)
  if (clean.length > 10) return false;

  // Valid aircraft pattern
  return VALID_AIRCRAFT.test(clean);
}

// Parse all months
const monthFolders = fs.readdirSync(STATS_DIR)
  .filter(f => fs.statSync(path.join(STATS_DIR, f)).isDirectory())
  .sort();

console.log(`📂 Found ${monthFolders.length} months\n`);

for (const monthFolder of monthFolders) {
  const monthPath = path.join(STATS_DIR, monthFolder);

  // Find Excel file
  const files = fs.readdirSync(monthPath)
    .filter(f => f.includes('Dnevni izvještaj o saobraćaju') && f.endsWith('.xlsx'));

  if (files.length === 0) continue;

  const filePath = path.join(monthPath, files[0]);

  try {
    console.log(`📊 ${monthFolder.substring(0, 10)}: `, );

    const workbook = XLSX.readFile(filePath);
    let totalFlights = 0;

    // Iterate through all sheets (days)
    for (const sheetName of workbook.SheetNames) {
      const worksheet = workbook.Sheets[sheetName];
      const data = XLSX.utils.sheet_to_json(worksheet, { header: 1 }) as any[][];

      // Skip header
      for (let i = 1; i < data.length; i++) {
        const row = data[i];

        // Skip empty rows
        if (!row[0]) continue;

        totalFlights++;

        // Extract data
        // [0] datum, [1] kompanija, [2] ruta, [3] tip a/c, [4] reg, [5] tip OPER
        const airline = row[1];
        const aircraft = row[3];
        const opType = row[5];

        // Airline
        if (airline && typeof airline === 'string') {
          const normalized = normalizeAirline(airline);
          airlines.set(normalized, (airlines.get(normalized) || 0) + 1);
        }

        // Aircraft type
        if (aircraft && typeof aircraft === 'string') {
          const clean = aircraft.trim();
          if (isValidAircraft(clean)) {
            aircraftTypes.set(clean, (aircraftTypes.get(clean) || 0) + 1);
          }
        }

        // Operation type
        if (opType && typeof opType === 'string') {
          const clean = opType.trim().toUpperCase();
          // Skip if it's a number (capacity)
          if (!/^\d+$/.test(clean)) {
            operationTypes.set(clean, (operationTypes.get(clean) || 0) + 1);
          }
        }
      }
    }

    console.log(`✅ ${totalFlights} flights`);

  } catch (error: any) {
    console.log(`❌ Error: ${error.message}`);
  }
}

// Sort by count
const sortedAirlines = Array.from(airlines.entries())
  .sort((a, b) => b[1] - a[1]);

const sortedAircraft = Array.from(aircraftTypes.entries())
  .sort((a, b) => b[1] - a[1]);

const sortedOpTypes = Array.from(operationTypes.entries())
  .sort((a, b) => b[1] - a[1]);

// Print results
console.log('\n' + '='.repeat(80));
console.log('📊 EKSTRAKTOVANI PODACI IZ 2025 GODINE (CLEAN)');
console.log('='.repeat(80));

console.log(`\n✈️  AVIOKOMPANIJE (${sortedAirlines.length}):`);
sortedAirlines.forEach(([name, count]) => {
  console.log(`  ${count.toString().padStart(4)} x ${name}`);
});

console.log(`\n🛩️  TIPOVI AVIONA (${sortedAircraft.length}):`);
sortedAircraft.forEach(([model, count]) => {
  console.log(`  ${count.toString().padStart(4)} x ${model}`);
});

console.log(`\n🔧 TIPOVI OPERACIJA (${sortedOpTypes.length}):`);
sortedOpTypes.forEach(([type, count]) => {
  console.log(`  ${count.toString().padStart(4)} x ${type}`);
});

// Save to JSON
const output = {
  airlines: sortedAirlines.map(([name, count]) => ({ name, count })),
  aircraftTypes: sortedAircraft.map(([model, count]) => ({ model, count })),
  operationTypes: sortedOpTypes.map(([type, count]) => ({ type, count })),
};

fs.writeFileSync(
  'output/2025-extracted-data.json',
  JSON.stringify(output, null, 2)
);

console.log('\n✅ Data saved to: output/2025-extracted-data.json');
